Client/src/App/DataRestriction/DataRestrictionActionCreators.ts

```diff
@@ -0,0 +1,95 @@
+import { ActionCreatorServices } from '@veupathdb/wdk-client/lib/Core/WdkMiddleware';
+import { makeActionCreator, InferAction } from '@veupathdb/wdk-client/lib/Utils/ActionCreatorUtils';
+
+import { fetchStudies } from 'ebrc-client/App/Studies/StudyActionCreators';
+import { checkPermissions, UserPermissions } from 'ebrc-client/StudyAccess/permission';
+
+import { isAllowedAccess } from './DataRestrictionUtils';
+
+export type Action =
+  | InferAction<typeof restricted>
+  | InferAction<typeof unrestricted>
+  | InferAction<typeof clearRestrictions>;
+
+// FIXME: Retire this type once the study entities derived by fetchStudies
+// have been properly typed
+export type Study = any;
+
+// FIXME: Retire this type once DataRestrictionUtils has been properly typed
+export type DataRestrictionActionType = string;
+
+export interface ActionAttemptDetails {
+  studyId: string;
+  onAllow?: () => void;
+  onDeny?: () => void;
+}
+
+export function attemptAction(action: DataRestrictionActionType, details: ActionAttemptDetails) {
```

Comment:
I wanted to type this action creator by annotating its return type as `ActionThunk<Action>`, but that broke the build. I believe the culprit is [this unresolved bug regarding the use of certain recursive types and async-await](https://github.com/microsoft/TypeScript/issues/35289).

(Ironically, the reporter of that bug also seemed to have been working on some kind of redux / state management with a recursive action type!)

Reply:
Very interesting!